File: rest-service/src/main/java/mb/lib/queue/model/CreateRequest.java
package mb.lib.queue.model;

import com.fasterxml.jackson.annotation.JsonGetter;

public class CreateRequest <T>
{
  private final String host;
  private final T payload;

  public CreateRequest(String host, T payload) {
    this.host    = host;
    this.payload = payload;
  }

  @JsonGetter("url")
  public String getHost() {
    return host;
  }

  @JsonGetter("payload")
  public T getPayload() {
    return payload;
  }
}

File: rest-service/src/main/kotlin/mb/lib/queue/model/CreateRequest.kt
package mb.lib.queue.model

import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.JsonNode
import org.veupathdb.lib.fireworq.JobRequest
import org.veupathdb.lib.jackson.Json

open class CreateRequest<T>(val host: String, val payload: T) {
  @JsonValue
  fun toJSON(): JsonNode = Json.newObject {
    put("url", host)
    putPOJO("payload", payload)
  }

  fun toInternal() = JobRequest(host, payload)
}
